packages/grafana-ui/src/components/Combobox/MultiCombobox.tsx

```diff
@@ -45,17 +50,38 @@ export const MultiCombobox = <T extends string | number>(props: MultiComboboxPro
   const styles = useStyles2(getComboboxStyles);
   const [inputValue, setInputValue] = useState('');
 
-  const [baseItems, baseSetItems] = useState(isAsync ? [] : options);
+  const allOptionItem = useMemo(() => {
+    return {
+      label:
+        inputValue === ''
+          ? t('multicombobox.all.title', 'All')
+          : t('multicombobox.all.title-filtered', 'All (filtered)'),
+      // Type casting needed to make this work when T is a number
+      value: ALL_OPTION_VALUE,
+    } as ComboboxOption<T>;
+  }, [inputValue]);
+  const getOptionsToSet = useCallback(() => {
+    return isAsync ? [] : enableAllOption ? [allOptionItem, ...options] : options;
+  }, [options, enableAllOption, allOptionItem, isAsync]);
 
-  const items = useMemo(() => baseItems.filter(itemFilter(inputValue)), [baseItems, inputValue]);
+  const [baseItems, baseSetItems] = useState(getOptionsToSet());
 
-  // TODO: Improve this with async
-  useEffect(() => {
-    baseSetItems(isAsync ? [] : options);
-  }, [options, isAsync]);
+  const items = useMemo(() => {
+    const newItems = baseItems.filter(itemFilter(inputValue));
 
+    if (enableAllOption && newItems.length === 1 && newItems[0] === allOptionItem) {
+      return [];
+    }
+
+    return newItems;
+  }, [baseItems, inputValue, enableAllOption, allOptionItem]);
   const [isOpen, setIsOpen] = useState(false);
 
+  // TODO: Improve this with async
+  useEffect(() => {
+    baseSetItems(getOptionsToSet());
```

Comment:
Can we change this to a useMemo instead?

Reply:
For the sake of learning, could you say why use `useMemo` here instead of `useEffect`? To cache the result? 🤔

Reply:
One of the reasons is because `baseSetItems` is only used here, and nowhere else. The logic here is that we set a value (the state) when some of the dependencies change. In this case we can just remove the state for base items and replac just change `getOptionsToSet` to a memoized `baseItems` instead. So basically, we just remove a bunch of unneccsary code here.

Generally, `useEffect` updates the following render, which may cause inconsistencies. `useEffect` should be used for addressing side effects, such as requests, event listeners or similar behavior. If there is a `set` function in the dependencies, that thing should probably be using `useMemo` instead.

Reply:
I think the react docs have a simple example of why it can be problematic https://react.dev/learn/you-might-not-need-an-effect#updating-state-based-on-props-or-state

But basically, like Tobias said, using useEffect to sync a prop to internal state can lead to annoying out-of-sync issues and has worse performance.

Tobias said it great that useEffect should be for (side) effects external to the component.

Reply:
Thank you both. Tbh, I missed it was related to state. I know `useEffect` + state is not a good combination at all. My bad. Sometimes, I find it hard to decide whether I should use `useMemo` or not, and I thought it was a good opportunity to ask, but I totally messed it up. Again, thanks 🙂